### DISCUSSION TITLE
Processing multiple channels

### DISCUSSION THREAD

**Discussion by J4NPM** [neuropsychology/NeuroKit#519] [Q&A]
Hi there,

Congratulations for everyone involved in developing such a great project!
Is there a way to process multiple simultaneous channels at once, such as 2 ECG channels in a 2D Array (samples, channels) as input for ecg_process(), ecg_peaks() and other functions?

Thank you very much

**Comment by zen-juen** [comment]
Hi @J4NPM, firstly thanks for the congratulations 😄 

Both `ecg_process()` and `ecg_peaks()` (as well as other signal-specific functions) take only one ECG channel as the default input, so you'll probably need to loop these functions through each channel if you've got 2 ECG channels in one data structure.

However, if you have an n-dimensional data structure containing channels of *different* biosignals such as like so:
```
data
Out[4]: 
            ECG        EDA  Photosensor       RSP
0     -0.015869  13.196868          5.0  0.778931
1     -0.011703  13.197173          5.0  0.777588
2     -0.009766  13.197020          5.0  0.777435
3     -0.013321  13.197631          5.0  0.777557
4     -0.009583  13.196715          5.0  0.775299
        ...        ...          ...       ...
14995 -0.007675  14.873656          5.0  1.351868
14996 -0.005203  14.873961          5.0  1.356689
14997 -0.004272  14.874266          5.0  1.361938
14998 -0.000458  14.875335          5.0  1.366821
14999  0.000244  14.874114          5.0  1.372040

[15000 rows x 4 columns]
```
You can pass this through our `bio_process()` function which is able to process multiple channels at once by specifying with column corresponds to which biosignal.
```
df, info = nk.bio_process(ecg=data["ECG"], rsp=data["RSP"], eda=data["EDA"], keep=data["Photosensor"], sampling_rate=100)
``` 
Hope this helps!

**Comment by DominiqueMakowski** [comment]
I'd also add that we don't support multi-threading and parallel processing within neurokit, and we currently don't have plans to. That said, we're still looking forward to contributions that would make our code faster and more efficient, so if you have any leads on that side feel free to let us know
